### ISSUE TITLE
Weak global reference table overflow on Android 7.0 and 7.1 only

### ISSUE THREAD

**Issue by jvv-aip** [xamarin/xamarin-android#1545]
We've received messages from a lot of our users running Android 7.0 and 7.1 devices that our app crashes randomly. 

After investigating  issue seems to be this error "JNI ERROR (app bug): weak global reference table overflow (max=51200)" resulting in 'tgkill' app crash reports in Google Play 

We do add some views in our code programmatically and given information from our users, it occurs after working with the app for a while on random moments.

I've tried reproducing this issue by making a project in which I can reproduce this issue.
In this code I add 5000 TextViews in a loop to a LinearLayout after clicking a button. Before starting the loop I call  RemoveAllViews on the LinearLayout.
This works on all Android versions I've tested (4.4, 5.1, 60, 8.1) but on Android 7.0 and 7.1 the 5th time I click this button.
In "Steps to Reproduce" I've attached the repro project with steps to reproduce.

I've also contacted Microsoft support about this, with as last insight: 
"We investigated this further now and it seems that the native implementation (which depends on the OS) is different for the controls on Android 7,
which might explain why it’s not Disposed correctly in Java.

Actually Xamarin creates OS-specific controls; as I understand it’s up to Android to handle the Java stack correctly.
The different results we see on different Android OS versions might depend on implementation differences between the Android versions,  or it might depend on the OS-specific Xamarin controls
"
After this message I was given the advise to create a bug report here.

### Steps to Reproduce

[sampleproject.zip](https://github.com/xamarin/xamarin-android/files/1902054/sampleproject.zip)

1. Open the attached sample project, run on an Android 7 or 7.1 device (not on emulator because this has a lower weak global references limit)
2. Click the add views button 4 times
3. This works
4. Click it the fifth time, now it crashes due to a global ref table overflow

5. Run the same code on another Android version (I've tested on 4.4, 5.1, 6.0 and 8.1)
6. The app does not crash after clicking the button the fifth time.

### Expected Behavior
When nearing the weak global ref limit, the garbage collector should clean this, like it does on other Android versions
### Actual Behavior
It doesn't clean up, so a weak global ref table overflow error occurs

### Version Information

=== Visual Studio Enterprise 2017 for Mac ===

Version 7.4.2 (build 12)
Installation UUID: e3620547-d246-4b4c-a6e4-6560e88cc3ec
Runtime:
	Mono 5.8.1.0 (2017-10/6bf3922f3fd) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)

	Package version: 508010000

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Version: 2.0.5
SDK: /usr/local/share/dotnet/sdk/2.1.4/Sdks
SDK Version: 2.1.4
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.8.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.1
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin.iOS ===

Version: 11.9.1.24 (Visual Studio Enterprise)
Hash: f62de472
Branch: xcode9.3
Build date: 2018-03-29 19:30:53-0400

=== Xamarin.Android ===

Version: 8.2.0.16 (Visual Studio Enterprise)
Android SDK: /Users/jordyvanvijfeijken/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		4.0.3 (API level 15)
		4.1   (API level 16)
		4.2   (API level 17)
		4.3   (API level 18)
		4.4   (API level 19)
		6.0   (API level 23)
		7.0   (API level 24)
		7.1   (API level 25)
		8.0   (API level 26)
		8.1   (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 27.0.1
SDK Build Tools Version: 27.0.3

Java SDK: /usr
java version "1.8.0_111"
Java(TM) SE Runtime Environment (build 1.8.0_111-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.111-b14, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 704020012
Git revision: 0d8e3f0a4d683771f17959739956fa09c7ba21e3
Build date: 2018-03-30 10:45:17-04
Xamarin addins: 958839ea56ab1e331caf7c92b6ad50fb9e6ee9d2
Build lane: monodevelop-lion-d15-6

=== Operating System ===

Mac OS X 10.13.3
Darwin 17.4.0 Darwin Kernel Version 17.4.0
    Sun Dec 17 09:19:54 PST 2017
    root:xnu-4570.41.2~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Open With 0.1

### Log File

[logfile.txt](https://github.com/xamarin/xamarin-android/files/1902084/logfile.txt)

**Comment by jonpryor** [comment]
You are unfortunately running into the limits of Xamarin.Android. If possible, please back away from those limits.

TL;DR: If at all possible, avoid architectures which involve adding *thousands* of *anything* to a `LinearLayout`, and instead use *Adapters* such as [`ListAdapter`](https://docs.microsoft.com/en-us/xamarin/android/user-interface/layouts/list-view/populating) so that you don't need to have a UI element for everything in your list, but instead have a UI element for *only the items shown on the screen*.

Alternatively, see if you can [help the GC](https://docs.microsoft.com/en-us/xamarin/android/internals/garbage-collection#helping-the-gc) by disposing of known Java-only types once you're done with them. For example, in `MainActivity.AddViews()` (in the sample project):

```csharp
void AddViews()
{
    _llContent.RemoveAllViews();
    for (int i = 0; i < viewCount; i++)
    {
        using (var textView = new TextView(this)) {
            textView.Text = $"Text line {count} - {i + 1}";
            textView.SetTextColor(Color.White);
            textView.SetTextSize(Android.Util.ComplexUnitType.Sp, 16.0f);
            _llContent.AddView(textView);
            using (var layoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MatchParent, ViewGroup.LayoutParams.WrapContent)) {
                layoutParams.SetMargins(5, 5, 5, 5);
                textView.LayoutParameters = layoutParams;
            }
        }
    }
}
```

**Background**: 

[Cross-VM garbage collection](https://docs.microsoft.com/en-us/xamarin/android/internals/garbage-collection#cross-vm-object-collections) involves the following steps:

1. All Xamarin values *eligible for collection by Mono's GC* have their JNI Global Reference replaced with a JNI Weak global reference.
2. The Java-side GC is invoked.
3. We attempt to create JNI Global References from the JNI Weak Global References the values from (1). If a Global Reference can be created, the value is still alive and can't be collected; otherwise, the Xamarin value is garbage, and will be collected by the GC.

([Relevant source code](https://github.com/xamarin/xamarin-android/blob/master/src/monodroid/jni/monodroid-glue.c#L1136-L1657))

Additionally, Android only allow creating ~51200 JNI Global References, and 51200 JNI Weak Global References, and *every* `Java.Lang.Object` subclass uses one of those 51200 values.

**Analysis**

In short, you can create *lots* of Java.Lang.Object subclass values, but you can't go *too* crazy. Adding 5000 `TextView` instances on a button press is bordering on "too crazy" territory. And yes, that's presumably "just" for your sample app, but if at all possible "normal" app use should stay *far away* from the upper Global Reference limits, and the app should be architected to avoid those limits (e.g. by using `Adapter<T>`).

For example, `logfile.txt` contains:

```
5955 of android.widget.LinearLayout (5955 unique instances)
2275 of md5eb12cd8826d26d24f62cd8a09277051b.ObservationOverviewPropertyValueViewHolder (2275 unique instances)
 730 of android.support.v7.widget.AppCompatTextView (730 unique instances)
 259 of android.support.v7.widget.AppCompatImageView (259 unique instances
 228 of mono.android.view.View_OnClickListenerImplementor (228 unique instances)
 204 of android.support.v7.widget.LinearLayoutManager (204 unique instances)
```

I see there are some types with `Adapter` in the name, but 5955 `LinearLayout` instances?! Ouch.

That said, the fact that this *only* happens on Android 7.0 and 7.1 kinda smells like an Android bug. Xamarin.Android *always* creates JNI Global References, and only uses Weak Global References during GC operations. Thus, in "normal" use, you'd create *X* Global References, a GC would occur, and (at most!) *X* Weak Global References would be created. I don't immediately see how you could have < 51200 Global References -- which *must* be the case, otherwise Android would have aborted the process *before* we started a GC! -- and wind up with *more* than 51200 Weak Global References being created; that doesn't make sense to me.

---

Looking back at the original report:

> **Actual Behavior**  
> It doesn't clean up, so a weak global ref table overflow error occurs

In point of fact, *it's trying to clean up*. The fact that Weak Global References are being created *at all* means that a GC is being performed, and it's through a GC that "clean up" is performed.

**Comment by jvv-aip** [comment]
We're using an Adapter to add the views to the RecylerView with an override of onBindViewHolder where the actual binding happens.

I'll try to trigger a GC with NotifyDataSetChanged() and see if that fixes something.
